batch/batch/front_end/query/operators.py

```diff
@@ -0,0 +1,121 @@
+import abc
+import re
+
+from ...exceptions import QueryError
+
+
+class Operator(abc.ABC):
+    regex: re.Pattern
+
+    @abc.abstractmethod
+    def to_sql(self) -> str:
+        raise NotImplementedError
+
+
+class ComparisonOperator(Operator, abc.ABC):
+    symbols = {'>=', '>', '<', '<=', '==', '=', '!='}
+
+
+class MatchOperator(Operator, abc.ABC):
+    symbols = {'=', '!=', '!~', '=~'}
+
+
+class PartialMatchOperator(MatchOperator, abc.ABC):
+    pass
+
+
+class ExactMatchOperator(MatchOperator, abc.ABC):
+    pass
+
+
+class GreaterThanEqualOperator(ComparisonOperator):
+    regex = re.compile('([^>]+)(>=)(.+)')
+
+    def to_sql(self) -> str:
+        return '>='
+
+
+class LessThanEqualOperator(ComparisonOperator):
+    regex = re.compile('([^<]+)(<=)(.+)')
+
+    def to_sql(self) -> str:
+        return '<='
+
+
+class GreaterThanOperator(ComparisonOperator):
+    regex = re.compile('([^>]+)(>)(.+)')
+
+    def to_sql(self) -> str:
+        return '>'
+
+
+class LessThanOperator(ComparisonOperator):
+    regex = re.compile('([^<]+)(<)(.+)')
+
+    def to_sql(self) -> str:
+        return '<'
+
+
+class NotEqualExactMatchOperator(ExactMatchOperator, ComparisonOperator):
+    regex = re.compile('([^!]+)(!=)(.+)')
+
+    def to_sql(self) -> str:
+        return '!='
+
+
+class EqualExactMatchOperator(ExactMatchOperator, ComparisonOperator):
+    regex = re.compile('([^=]+)(==|=)(.+)')
+
+    def to_sql(self) -> str:
+        return '='
+
+
+class NotEqualPartialMatchOperator(PartialMatchOperator):
+    regex = re.compile('([^!]+)(!~)(.+)')
+
+    def to_sql(self) -> str:
+        return 'NOT LIKE'
+
+
+class EqualPartialMatchOperator(PartialMatchOperator):
+    regex = re.compile('([^=]+)(=~)(.+)')
+
+    def to_sql(self) -> str:
+        return 'LIKE'
+
+
+symbols_to_operator = {
+    '>=': GreaterThanEqualOperator(),
+    '<=': LessThanEqualOperator(),
+    '>': GreaterThanOperator(),
+    '<': LessThanOperator(),
+    '!~': NotEqualPartialMatchOperator(),
+    '=~': EqualPartialMatchOperator(),
+    '!=': NotEqualExactMatchOperator(),
+    '==': EqualExactMatchOperator(),
+    '=': EqualExactMatchOperator(),
+}
+
+
+def pad_operator(s: str) -> str:
+    for op in symbols_to_operator.values():
+        match = op.regex.fullmatch(s)
+        if match:
+            groups = match.groups()
+            if len(groups) != 3:
+                raise QueryError(f'could not parse operator in query string "{s}"')
+            return ' '.join(groups)
+    raise QueryError(f'could not parse operator in query string "{s}"')
+
+
+def pad_maybe_operator(s: str) -> str:
+    for symbol in symbols_to_operator:
+        if symbol in s:
+            return pad_operator(s)
+    return s
```

Comment:
Ah I see this is for supporting operators without whitespace padding. That makes sense to suppor,t but it feels a bit complicated to find out the operator, then use a regular expression to pad the string, then parse it again later. What if instead of `return pad_operator(s)` we returned `[token.strip() for token in s.split(symbol)]` and this function instead of returning a `str` returned a `List[str]`. That basically cuts through straight to the point of having split on whitespace and removes the need for any regular expression parsing.

Reply:
I needed the actual operator in the result string which your example doesn't do. Not sure I ended up with a better solution, but probably nicer than what it was before.

Reply:
I'm ok with this. We should also be able to delete all the regular expression stuff right?